### PR TITLE
Add multi-stage build of docker edge image

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Add multi-stage build of docker edge image

## Motivation and Context
This is to reduce the overall size of the image,
we only retain the binary that was built in previous stage.

## Regression
No

## How Has This Been Tested?
Current docker edge image is 

```
minio/minio                        edge                5519a6f2d1fe        7 hours ago         483MB
```

New docker image will be 
```
minio/minio                        edge                8b5c246055c6        9 seconds ago       67.6MB
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by nitisht — DISMISSED** [review]
LGTM & Tested

File: Dockerfile [hunk 1]
```diff
@@ -1,28 +1,36 @@
...
      apk add --no-cache ca-certificates 'curl>7.61.0' && \
-     apk add --no-cache --virtual .build-deps git && \
      echo 'hosts: files mdns4_minimal [NOTFOUND=return] dns mdns4' >> /etc/nsswitch.conf && \
-     go get -v -d github.com/minio/minio && \
-     cd /go/src/github.com/minio/minio && \
-     go install -v -ldflags "$(go run buildscripts/gen-ldflags.go)" && \
-     rm -rf /go/pkg /go/src /usr/local/go && apk del .build-deps
-
-EXPOSE 9000
+     chmod +x /usr/bin/minio  && \
```
Comment (wlan0) [inline Dockerfile]: Why is this needed? I would think the filemode is set on the file on `go install`

File: Dockerfile [hunk 2]
```diff
@@ -1,28 +1,36 @@
...
+     chmod +x /usr/bin/docker-entrypoint.sh && \
```
Comment (wlan0) [inline Dockerfile]: see above comment

File: Dockerfile [hunk 3]
```diff
@@ -1,28 +1,36 @@
...
+     chmod +x /usr/bin/healthcheck.sh
```
Comment (wlan0) [inline Dockerfile]: see above comment

File: Dockerfile [hunk 4]
```diff
@@ -1,28 +1,36 @@
...
 LABEL maintainer="Minio Inc <dev@minio.io>"
 
 ENV GOPATH /go
-ENV PATH $PATH:$GOPATH/bin
 ENV CGO_ENABLED 0
+
+WORKDIR /go/src/github.com/minio/
+
+RUN  \
+     apk add --no-cache ca-certificates 'curl>7.61.0' git && \
```
Comment (wlan0) [inline Dockerfile]: This line is not necessary

File: Dockerfile [hunk 5]
```diff
@@ -1,28 +1,36 @@
...
+     echo 'hosts: files mdns4_minimal [NOTFOUND=return] dns mdns4' >> /etc/nsswitch.conf && \
```
Comment (wlan0) [inline Dockerfile]: This line is also unnecessary. We can just `go get` and then `go build/install`. Any particular reasons for the other packages? 

This is a `golang` image.

**Comment by wlan0** [comment]
This image pulls the latest (i guess that's why you call it `edge`). Is there a versioned `Dockerfile` for releases?

File: Dockerfile (hunk 1 above)
Comment (harshavardhana) [reply Dockerfile]: it is generally to ensure the right permissions, we do not know if COPY preserves or not.

File: Dockerfile (hunk 2 above)
Comment (harshavardhana) [reply Dockerfile]: Same as above

File: Dockerfile (hunk 3 above)
Comment (harshavardhana) [reply Dockerfile]: Same as above

File: Dockerfile (hunk 4 above)
Comment (harshavardhana) [reply Dockerfile]: Primarily for https `go get` without ca-certificates there is a possibility of breakage, `curl` might not be needed.

File: Dockerfile (hunk 5 above)
Comment (harshavardhana) [reply Dockerfile]: This might not needed will remove it.

**Comment by harshavardhana** [comment]
> This image pulls the latest (i guess that's why you call it `edge`). Is there a versioned `Dockerfile` for releases?

That is automatically done in Docker hub upon a release tag @wlan0

File: Dockerfile (hunk 1 above)
Comment (wlan0) [reply Dockerfile]: COPY preserves the filemode

File: Dockerfile (hunk 4 above)
Comment (wlan0) [reply Dockerfile]: You only need to install `git`. The image has certs. 

> curl might not be needed.

:+1:

Comment (harshavardhana) [reply Dockerfile]: Yeah that is pretty pathetic, go needs `git` and Golang image doesn't have it. That is why we kept everything that we need :-)

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

File: Dockerfile (hunk 2 above)
Comment (wlan0) [reply Dockerfile]: This line can also be removed

File: Dockerfile (hunk 3 above)
Comment (wlan0) [reply Dockerfile]: Please remove this line as well

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by wlan0 — APPROVED** [review]

**Review by wlan0 — APPROVED** [review]

**Comment by wlan0** [comment]
LGTM

**Review by nitisht — APPROVED** [review]
